### ISSUE TITLE
ZMQ.REQ regression since 0.4.0

### ISSUE THREAD

**Issue by zezke** [zeromq/jeromq#497]
I am writing a library that communicates with the ZeroMQ server of OpenNMT. This code works using 0.3.6, but hangs on the receiving using any 0.4.x branch.

```java
    public OpenNMTZeroMQClient(String host, int port) {
        this.context = ZMQ.context(2);
        this.socket = context.socket(ZMQ.REQ);
        String hostAddress = "tcp://" + host + ":" + port;
        LOG.info("Connecting to '{}'", hostAddress);
        socket.connect(hostAddress);
    }


    public void translate(String text) {
        String js = new JSONArray()
                .put(new JSONObject()
                        .put("src", text))
                .toString();
        LOG.info("Sending '{}'...", js);
        boolean wasSent = socket.send(js.getBytes(ZMQ.CHARSET), 0);
        if(wasSent) {
            LOG.info("Sent!");
        }
        LOG.info("Receiving...");
        String reply = socket.recvStr();
        LOG.info("Received: '{}'", reply);
    }

    public void close() {
        if(this.socket != null) {
            this.socket.close();
        }
        if(this.context != null) {
            this.context.term();
        }
    }
```

It doesn't actually fail, but it just hangs on the recv() call. The server didn't receive anything, so the problem is probably that the send never happened.

EDIT: code formatting

**Comment by trevorbernard** [comment]
How does it fail?

**Comment by daveyarwood** [comment]
Silly question: could it be that the requests are being sent to the wrong host or port?

Simple REQ/REP is a basic use case for ZeroMQ -- it ought to work. Anecdotally, I can confirm that I am able to use JeroMQ 0.4.2 to run both a server and client (in separate processes) and send/receive messages back and forth in a loop.

**Comment by zezke** [comment]
REQ/REP is indeed working for 0.4.2 if I write both the server and client and run them in a separate thread. For some reason it does not connect/communicate with the remote host.

Changing the Maven import to 0.3.6 instantly fixes the behavior.

**Comment by daveyarwood** [comment]
Oh, right. That sounds related to https://github.com/zeromq/jeromq/issues/466

I think we're just waiting on an 0.4.3 release.

**Comment by trevorbernard** [comment]
@daveyarwood @fredoboulo a bit off topic, but what are your thoughts on releasing 0.4.3? Are there any issues you'd like to stabilize first? Or is it fine?

**Comment by daveyarwood** [comment]
It seems like 0.4.1 and 0.4.2 are not very stable, and there is definitely the issue where those releases are not compatible with ZMQ 3.X, which to me is the most pressing issue. I would advocate for releasing 0.4.3 and seeing which recent issues we can resolve.

I'll defer to @fredoboulo about any outstanding issues that might block release... there are none that I'm aware of.

**Comment by fredoboulo** [comment]
Yes, it sounds related to protocol compatibility.

0.4.1 and 0.4.2 are definitely not stable, a lot of fixes occurred to stabilize the situation already, I would advocate for release as soon as possible, with one extra test: check the reconnection automagically happens when network is down. It's one of the features that cannot be simply tested.

I have only one matter in the pipeline: #495 , to provide a way to close the poller. I made a commit but no PR, maybe release could happen after that (matter of days if not hours)?

How does that sound to you, @trevorbernard @daveyarwood ?

**Comment by daveyarwood** [comment]
I took a look at your commit and it looks good for me. I think we can merge that in and then make an 0.4.3 release.

**Comment by daveyarwood** [comment]
PR merged. I think we are clear for an 0.4.3 release.

**Comment by fredoboulo** [comment]
Please test this reconnection before we release binary jar

**Comment by fredoboulo** [comment]
So .... this is not strictly related to this issue; but reconnection is tested on my side (ubuntu 16.04; java 1.8.0_131). Tested by using LocalThr and RemoteThr with infinite message sending, and switching the wifi off then on again

**Comment by zezke** [comment]
I am observing the same behavior with 0.4.3-SNAPSHOT.

**Comment by fredoboulo** [comment]
@trevorbernard I propose in that case to postpone next release until the situation of this issue is clarified, we got two different feedbacks:
[0.3.6 REQ <-> 0.4.3-SNAPSHOT REP working](https://github.com/zeromq/jeromq/issues/466#issuecomment-337140483)

0.4.3-SNAPSHOT REQ <-> 0.3.6 REP broken (that's my understanding of this issue)

**Comment by fredoboulo** [comment]
@zezke which version of zeromq does the other side uses?

**Comment by zezke** [comment]
The server is running:

- Ubuntu 16.04
- libzmq-dev 2.2.0+dfsg-7
- lua-zmq 1.1.1

**Comment by fredoboulo** [comment]
Hi @zezke, I think  have found the root cause, could you try with the commit I referenced here? If that fixes the situation, I will make a PR

**Comment by fredoboulo** [comment]
This issue might have been closed automatically by Github. @zezke, please check with latest SNAPSHOT if the issue is still occurring.

**Comment by trevorbernard** [comment]
@fredoboulo @daveyarwood Seems like it's a good time to release? Any way I can reach you other than github? My email is on my profile.

**Comment by daveyarwood** [comment]
Seems good to me. My email is on my profile too, definitely feel free to email me!

**Comment by zezke** [comment]
@fredoboulo The 0.4.3 release does indeed fix the issue. Thanks a bunch and sorry for the delay on my end!
